executor/src/witgen/range_constraints.rs

```diff
@@ -194,50 +194,69 @@ fn range_multiple<T: FieldElement>(min: T, max: T, factor: T) -> (T, T) {
 /// There are cases where the intersection cannot be represented as a single internal.
 /// in that case, it returns the smaller of the two inputs.
 /// If the intersection is empty, returns None.
-fn interval_intersection<T: FieldElement>(mut a: (T, T), mut b: (T, T)) -> Option<(T, T)> {
+fn interval_intersection<T: FieldElement>(a: (T, T), b: (T, T)) -> Option<(T, T)> {
     // We shift both intervals until they are both non-wrapping intervals.
     // If we do not succeed after shifting both of them by the smallest amount,
     // it means that the intersection cannot be expressed as a single interval.
     // In that case we just choose the smaller of the two inputs.
-    let shift: T = if a.0 <= a.1 {
-        // a is regular
-        if b.0 <= b.1 {
-            0.into()
-        } else if a.1 <= b.1 || b.1 <= a.0 {
-            b.1
-        } else if a.1 < b.0 {
-            a.1
-        } else {
-            // intersection consists of two intervals.
-            if range_width(a.0, a.1) >= range_width(b.0, b.1) {
-                return Some(a);
-            } else {
-                return Some(b);
-            }
-        }
-    } else if b.0 <= b.1 {
-        // a is wrapping b is regular
-        return interval_intersection(b, a);
+    // let shift: T = if a.0 <= a.1 {
+    //     // a is regular
+    //     if b.0 <= b.1 {
+    //         0.into()
+    //     } else if a.1 <= b.1 || b.1 <= a.0 {
+    //         b.1
+    //     } else if a.1 < b.0 {
+    //         a.1
+    //     } else {
+    //         // intersection consists of two intervals.
+    //         if range_width(a.0, a.1) >= range_width(b.0, b.1) {
+    //             return Some(a);
+    //         } else {
+    //             return Some(b);
+    //         }
+    //     }
+    // } else if b.0 <= b.1 {
+    //     // a is wrapping b is regular
+    //     return interval_intersection(b, a);
+    // } else {
+    //     // both a and b are wrapping
+    //     if cmp::max(a.1, b.1) < cmp::min(a.0, b.0) {
+    //         cmp::max(a.1, b.1)
+    //     } else {
+    //         // intersection consists of two intervals
+    //         if range_width(a.0, a.1) >= range_width(b.0, b.1) {
+    //             return Some(a);
+    //         } else {
+    //             return Some(b);
+    //         }
+    //     }
+    // };
+    if let Some((shift, (a_shifted, b_shifted))) =
+        [0.into(), a.0, a.1 + 1.into(), b.0, b.1 + 1.into()]
```

Comment:
`0.into()` is not really needed.

And neither `a.0` or `b.0`, it seems. I think they are redundant, because every non-union interval that might start at with `a.1 + 1` or `b.1 + 1` have to end in either `a.0 - 1` or `b.0 - 1`.

Also, since you are using `-shift` now, the correct here (even if redundant) would be `-a.0` and `-b.0`: range starts have to be shifted forward, range ends have to be shifted backwards minus one.

Or you do the opposite, use only `a.0` and `b.0`, and pass positive `shift` to `shifted_interval`.

Reply:
yeah, that's a good point. Weird why I had the other options in the complicated case distinction.